google-cloud-spanner/src/test/java/com/google/cloud/spanner/DatabaseClientImplTest.java

```diff
@@ -195,6 +198,179 @@ public void tearDown() {
+            .setMinSessions(1)
+            .setMaxSessions(1) // to ensure there is 1 session and pool is 100% utilized
+            .setCloseIfInactiveTransactions()
+            .setInactiveTransactionRemovalOptions(inactiveTransactionRemovalOptions)
+            .setLoopFrequency(1000L) // main thread runs every 1s
+            .build();
+    spanner =
+        SpannerOptions.newBuilder()
+            .setProjectId(TEST_PROJECT)
+            .setDatabaseRole(TEST_DATABASE_ROLE)
+            .setChannelProvider(channelProvider)
+            .setCredentials(NoCredentials.getInstance())
+            .setSessionPoolOption(sessionPoolOptions)
+            .build()
+            .getService();
+    DatabaseClientImpl client =
+        (DatabaseClientImpl)
+            spanner.getDatabaseClient(DatabaseId.of(TEST_PROJECT, TEST_INSTANCE, TEST_DATABASE));
+    Instant initialExecutionTime = client.pool.poolMaintainer.lastExecutionTime;
+
+    try (TransactionManager manager = client.transactionManager()) {
+      TransactionContext transaction = manager.begin();
+      mockSpanner.setCommitExecutionTime(
+          SimulatedExecutionTime.ofException(
+              mockSpanner.createSessionNotFoundException("TEST_SESSION_NAME")));
+      while (true) {
+        try {
+          // Simulate a delay of 4s to ensure that the below transaction is a long-running one.
+          // We require to wait for 4s so that the main thread executes at-least once every 1s
+          // As per this test, anything which takes more than 2s is long-running
+          mockSpanner.setExecuteSqlExecutionTime(
+              SimulatedExecutionTime.ofMinimumAndRandomTime(
+                  (int) Duration.ofSeconds(4).toMillis(), 0));
+          transaction.executeUpdate(UPDATE_STATEMENT);
+          manager.commit();
+          assertNotNull(manager.getCommitTimestamp());
+          break;
+        } catch (AbortedException e) {
+          transaction = manager.resetForRetry();
+        }
+        mockSpanner.setCommitExecutionTime(SimulatedExecutionTime.ofMinimumAndRandomTime(0, 0));
+      }
+    }
+    Instant endExecutionTime = client.pool.poolMaintainer.lastExecutionTime;
+
+    // first session executed update, session found to be long-running and cleaned up.
+    // During commit, SessionNotFound exception from backend caused replacement of session and
+    // transaction needs to be retried.
+    // On retry, session again found to be long-running and cleaned up.
+    // During commit, there was no exception from backend.
+    final int numSessionsInPool = client.pool.getNumberOfSessionsInPool();
+    final long numLeakedSessionsRemoved = client.pool.numLeakedSessionsRemoved();
+
+    assertNotEquals(
+        endExecutionTime,
+        initialExecutionTime); // if session clean up task runs then these timings won't match
+    assertTrue(numSessionsInPool > 0 && numSessionsInPool < 3);
+    assertThat(numLeakedSessionsRemoved > 0 && numLeakedSessionsRemoved <= 2);
```

Comment:
```suggestion
    assertTrue(numLeakedSessionsRemoved > 0 && numLeakedSessionsRemoved <= 2);
```

This is currently causing this test to falsely report that everything succeeded, as it is not really verifying anything. The expression in the `assertThat` call is actually false at the moment.